### PR TITLE
Support two outputs in split_reduce

### PR DESCRIPTION
**Pull request by pfultz2**

### CODE REVIEW

File: src/targets/gpu/hip.cpp [hunk 1]
```diff
@@ -304,9 +304,17 @@ argument get_preallocation(context& ctx, const std::string& id)
 
 void gpu_fill(context& ctx, const argument& dst, int value)
 {
-    // TODO: Handle non-packed tensor when value is not 0
-    assert(dst.get_shape().packed() and value == 0);
-    hip_async_memset(ctx, dst, value);
+    if(dst.get_sub_objects().empty())
+    {
+        // TODO: Handle non-packed tensor when value is not 0
+        assert(dst.get_shape().packed() and value == 0);
```
Comment (TedThemistokleous) [inline src/targets/gpu/hip.cpp:310]: Just assert here or do we want to add a throw if we hit that case but this isn't handled?

File: src/targets/gpu/kernels/include/migraphx/kernels/atomic.hpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+namespace migraphx {
+namespace atomic {
+
+using cas_rank = rank<1>;
+
+template <class T, class Op, MIGRAPHX_REQUIRES(sizeof(T) == 4 or sizeof(T) == 8)>
+MIGRAPHX_DEVICE_CONSTEXPR void cas(rank<1>, T& x, T y, Op op)
+{
+    MIGRAPHX_ATOMIC_CAS_WARNING();
+    using U    = conditional_t<sizeof(T) == 4, uint32_t, uint64_t>;
```
Comment (umangyadav) [inline src/targets/gpu/kernels/include/migraphx/kernels/atomic.hpp:31]: Tidy fixes required

**Review by TedThemistokleous — COMMENTED** [review]
Running this through CI again. I don't see anything outside the tidy errors.  Kicked off the other CI errors to see if there's still warnings with CK and ASAN

File: src/targets/gpu/hip.cpp (hunk 1 above)
Comment (pfultz2) [reply src/targets/gpu/hip.cpp:310]: Currently we are not calling this with a non-packed tensor which is why its an assert.

**Review by TedThemistokleous — APPROVED** [review]

File: src/targets/gpu/kernels/include/migraphx/kernels/rank.hpp [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#ifndef MIGRAPHX_GUARD_KERNELS_RANK_HPP
```
Comment (umangyadav) [inline src/targets/gpu/kernels/include/migraphx/kernels/rank.hpp:1]: License

File: src/targets/gpu/kernels/include/migraphx/kernels/atomic.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
+#ifndef MIGRAPHX_GUARD_KERNELS_ATOMIC_HPP
```
Comment (umangyadav) [inline src/targets/gpu/kernels/include/migraphx/kernels/atomic.hpp:1]: License

File: src/targets/gpu/kernels/include/migraphx/kernels/atomic.hpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+template <class T, class Op>
+MIGRAPHX_DEVICE_CONSTEXPR auto assign(rank<1>, T& x, T y, Op op)
+    MIGRAPHX_RETURNS(builtin_assign(x, y, op));
+
+} // namespace atomic
+
+template <class T, class U, class Op>
+MIGRAPHX_DEVICE_CONSTEXPR void atomic_assign(T& x, U y, Op op)
+{
+    atomic::assign(rank<1>{}, x, T(y), op);
```
Comment (umangyadav) [inline src/targets/gpu/kernels/include/migraphx/kernels/atomic.hpp:114]: Casting y to type T seems wrong. It should pick higher dtype between T and U

File: src/split_reduce.cpp [hunk 6]
```diff
@@ -162,19 +180,20 @@ void split_reduce::apply(module_pass_manager& mpm) const
...
-        auto splits = find_split(rm);
+        splitter s{rm};
+        auto splits = s.find_splits();
         if(splits.empty())
             continue;
         // Only use split reduce with float for now
         // TODO: Support half and other data types
         if(not std::all_of(splits.begin(), splits.end(), [](instruction_ref split) {
-               return split->get_shape().type() == shape::float_type;
+               return contains({shape::float_type, shape::half_type}, split->get_shape().type());
```
Comment (umangyadav) [inline src/split_reduce.cpp:190]: update TODO above to remove "half" type.

**Comment by umangyadav** [comment]
This PR fails to compile if i run
`MIGRAPHX_DISABLE_LAYERNORM_FUSION=1 MIGRAPHX_ENABLE_SPLIT_REDUCE=1  ./bin/test_verify "test_layernorm_large"`

**Comment by umangyadav** [comment]
This currently will not fuse 
`pointwise` preceding "reduce" ops is pointwise is alive after the reduce_ops. 

This is what happens with https://github.com/ROCm/AMDMIGraphX/issues/3212

```
@6 = convolution[padding={1, 1, 1, 1},stride={1, 1},dilation={1, 1},group=1,padding_mode=0](sample,@0) -> float_type, {2, 320, 64, 64}, {1310720, 4096, 64, 1}
@7 = broadcast[axis=1,out_lens={2, 320, 64, 64}](@1) -> float_type, {2, 320, 64, 64}, {0, 1, 0, 0}
@8 = contiguous(@7) -> float_type, {2, 320, 64, 64}, {1310720, 4096, 64, 1}
@9 = reshape[dims={2, 32, 10, 64, 64}](@6) -> float_type, {2, 32, 10, 64, 64}, {1310720, 40960, 4096, 64, 1}
@10 = reshape[dims={2, 32, 10, 64, 64}](@8) -> float_type, {2, 32, 10, 64, 64}, {1310720, 40960, 4096, 64, 1}
@11 = multibroadcast[out_lens={2, 320, 64, 64},out_dyn_dims={}](@3) -> float_type, {2, 320, 64, 64}, {0, 1, 0, 0}
@12 = contiguous(@11) -> float_type, {2, 320, 64, 64}, {1310720, 4096, 64, 1}
@13 = multibroadcast[out_lens={2, 320, 64, 64},out_dyn_dims={}](@4) -> float_type, {2, 320, 64, 64}, {0, 1, 0, 0}
@14 = contiguous(@13) -> float_type, {2, 320, 64, 64}, {1310720, 4096, 64, 1}
@15 = reshape[dims={2, 32, 10, 64, 64}](@12) -> float_type, {2, 32, 10, 64, 64}, {1310720, 40960, 4096, 64, 1}
@16 = reshape[dims={2, 32, 10, 64, 64}](@14) -> float_type, {2, 32, 10, 64, 64}, {1310720, 40960, 4096, 64, 1}
@17 = pointwise(@9,@10), [main:pointwise0] -> float_type, {2, 32, 10, 64, 64}, {1310720, 40960, 4096, 64, 1}
@18 = split_fused_reduce[axes={2, 3, 4},assign=assign_add](@17), [main:pointwise0:main:pointwise2:main:reduce_sum1:main:pointwise4:main:pointwise6:main:pointwise1:main:reduce_sum0_reshape_reshape:main:pointwise10_split] -> [float_type, {2, 32, 1, 1, 1}, {32, 1, 1, 1, 1}, float_type, {2, 32, 1, 1, 1}, {32, 1, 1, 1, 1}]
@19 = get_tuple_elem[index=0](@18) -> float_type, {2, 32, 1, 1, 1}, {32, 1, 1, 1, 1}
@20 = get_tuple_elem[index=1](@18) -> float_type, {2, 32, 1, 1, 1}, {32, 1, 1, 1, 1}
@21 = multibroadcast[out_lens={2, 32, 10, 64, 64},out_dyn_dims={}](@19) -> float_type, {2, 32, 10, 64, 64}, {32, 1, 0, 0, 0}
@22 = multibroadcast[out_lens={2, 32, 10, 64, 64},out_dyn_dims={}](@20) -> float_type, {2, 32, 10, 64, 64}, {32, 1, 0, 0, 0}
@23 = pointwise(@21,@22,@17,@21,@15,@16), [main:pointwise4] -> float_type, {2, 32, 10, 64, 64}, {1310720, 40960, 4096, 64, 1}
@24 = reshape[dims={2, 320, 64, 64}](@23) -> float_type, {2, 320, 64, 64}, {1310720, 4096, 64, 1}
@25 = convolution[padding={1, 1, 1, 1},stride={1, 1},dilation={1, 1},group=1,padding_mode=0](@24,@2) -> float_type, {2, 320, 64, 64}, {1310720, 4096, 64, 1}

```

@17 is not fused with @18 because @17 isused later at @23.

**Comment by pfultz2** [comment]
> @17 is not fused with @18 because @17 isused later at https://github.com/23.

Yea we need to support multi-output fusion in the future. We can probably just do the fusion initially with mlir.

**Comment by pfultz2** [comment]
> This PR fails to compile if i run `MIGRAPHX_DISABLE_LAYERNORM_FUSION=1 MIGRAPHX_ENABLE_SPLIT_REDUCE=1 ./bin/test_verify "test_layernorm_large"`

This is fixed.

File: src/split_reduce.cpp (hunk 6 above)
Comment (umangyadav) [reply src/split_reduce.cpp:190]: can you remove TODO for the Half ?

Comment (umangyadav) [reply src/split_reduce.cpp:190]: @pfultz2

File: test/split_reduce.cpp [hunk 7]
```diff
@@ -59,16 +59,24 @@ bool all_instructions_are_local(const migraphx::module& m)
...
 
+void auto_add_return(migraphx::module_ref m, migraphx::instruction_ref ins)
+{
+    m->add_return({ins});
+}
+
+void auto_add_return(migraphx::module_ref m, std::vector<migraphx::instruction_ref> inss)
+{
+    m->add_return(std::move(inss));
+}
```
[the remark this answers is not in the document]
Comment (umangyadav) [reply test/split_reduce.cpp:70]: @pfultz2  ^^ why do you require two overloads ?

File: src/split_reduce.cpp (hunk 6 above)
Comment (pfultz2) [reply src/split_reduce.cpp:190]: Updated.

File: test/split_reduce.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (umangyadav) [reply test/split_reduce.cpp:70]: It should construct `vector` arg automatically therefore you should only require one overload i guess

[the remark this answers is not in the document]
Comment (pfultz2) [reply test/split_reduce.cpp:70]: No it wont, there is no constructor in vector that takes a single `T`. Here is the compile errors without this function:

```
/data/test/split_reduce.cpp:88:20: error: no viable conversion from 'instruction_ref' (aka '_List_iterator<migraphx::instruction>') to 'std::vector<instruction_ref>' (aka 'vector<_List_iterator<migraphx::instruc
tion>>')
    rm->add_return(r);
                   ^
/data/test/split_reduce.cpp:115:16: note: in instantiation of function template specialization 'add_reduce_module<(lambda at /data/test/split_reduce.cpp:124:12)>' requested here
    auto rm  = add_reduce_module(p, name, inputs, axes, f);
               ^
/data/test/split_reduce.cpp:151:21: note: in instantiation of function template specialization 'add_reduce<(lambda at /data/test/split_reduce.cpp:124:12)>' requested here
        auto rsum = add_reduce(
                    ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:553:7: note: candidate constructor not viable: no known conversion from 'instruction_ref' (aka '_List_iterator<migraphx::instruction>
') to 'const vector<_List_iterator<instruction>> &' for 1st argument
      vector(const vector& __x)
      ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:572:7: note: candidate constructor not viable: no known conversion from 'instruction_ref' (aka '_List_iterator<migraphx::instruction>
') to 'vector<_List_iterator<instruction>> &&' for 1st argument
      vector(vector&&) noexcept = default;
      ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:625:7: note: candidate constructor not viable: no known conversion from 'instruction_ref' (aka '_List_iterator<migraphx::instruction>
') to 'initializer_list<value_type>' (aka 'initializer_list<std::_List_iterator<migraphx::instruction>>') for 1st argument
      vector(initializer_list<value_type> __l,
      ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:497:7: note: explicit constructor is not a candidate
      vector(const allocator_type& __a) _GLIBCXX_NOEXCEPT
      ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:510:7: note: explicit constructor is not a candidate
      vector(size_type __n, const allocator_type& __a = allocator_type())
      ^
/data/src/include/migraphx/module.hpp:177:61: note: passing argument to parameter 'args' here
    instruction_ref add_return(std::vector<instruction_ref> args);
                                                            ^
```

**Pull request merged**